content-src/lib/selectLayoutRender.js

```diff
@@ -23,8 +31,12 @@ export const selectLayoutRender = (state, prefs, rickRollCache) => {
           bufferRollCache.push(rickRoll);
         }
 
-        if (spocs.data.spocs[spocIndex] && rickRoll <= spocsConfig.probability) {
-          recommendations.splice(position.index, 0, spocs.data.spocs[spocIndex++]);
+        if (rickRoll <= spocsConfig.probability) {
+          spocIndex++;
+          recommendations.splice(position.index, 0, spoc);
+          chosenSpocs.add(spoc);
+        } else {
+          unchosenSpocs.add(spoc);
```

Comment:
Is it safe to assume unchosenSpocs set will always have size 1 as spocIndex will only increase if a spoc is chosen meaning a single spoc will use all positions if its not chosen. It might help to rename it to unchosenSpoc.

Reply:
Indeed, technically there should be only one item with the reason `probability_selection`. However, I think we do need a `Set` to record all the state changes for each SPOC item, as it might fail the selection for one position, then it succeeds for the next one. The same can occur to every single SPOC. We store all those changes here, then we will reconcile the final state for each SPOC at the end of this function. I've added some comments there to make this less confusing. Hopefully, that will help.

Reply:
Thanks that helps! [Wrote test ](https://gist.github.com/punamdahiya/db78f5f81f8ee1a9c6975330f4394548#file-spocsfill-unchosenspocs)to capture this case :)

Reply:
Nice test, will add it in.